x/oracle/abci.go

```diff
@@ -0,0 +1,135 @@
+package oracle
+
+import (
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/CosmosContracts/juno/v11/x/oracle/keeper"
+	"github.com/CosmosContracts/juno/v11/x/oracle/types"
+)
+
+// isPeriodLastBlock returns true if we are at the last block of the period
+func isPeriodLastBlock(ctx sdk.Context, blocksPerPeriod uint64) bool {
+	return (uint64(ctx.BlockHeight())+1)%blocksPerPeriod == 0
+}
+
+// EndBlocker is called at the end of every block
+func EndBlocker(ctx sdk.Context, k keeper.Keeper) error {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyEndBlocker)
+
+	params := k.GetParams(ctx)
+	if isPeriodLastBlock(ctx, params.VotePeriod) {
+		// Build claim map over all validators in active set
+		validatorClaimMap := make(map[string]types.Claim)
+		powerReduction := k.StakingKeeper.PowerReduction(ctx)
+		for _, v := range k.StakingKeeper.GetBondedValidatorsByPower(ctx) {
+			addr := v.GetOperator()
+			validatorClaimMap[addr.String()] = types.NewClaim(v.GetConsensusPower(powerReduction), 0, 0, addr)
+		}
+
+		var (
+			// voteTargets defines the symbol (ticker) denoms that we require votes on
+			voteTargets      []string
+			voteTargetDenoms []string
+		)
+		for _, v := range params.AcceptList {
+			voteTargets = append(voteTargets, v.SymbolDenom)
+			voteTargetDenoms = append(voteTargetDenoms, v.BaseDenom)
+		}
+
+		k.ClearExchangeRates(ctx)
+
+		// NOTE: it filters out inactive or jailed validators
+		ballotDenomSlice := k.OrganizeBallotByDenom(ctx, validatorClaimMap)
+
+		// Iterate through ballots and update exchange rates; drop if not enough votes have been achieved.
+		for _, ballotDenom := range ballotDenomSlice {
+			// Get weighted median of exchange rates
+			exchangeRate, err := Tally(ctx, ballotDenom.Ballot, params.RewardBand, validatorClaimMap)
+			if err != nil {
+				return err
+			}
+
+			// Set the exchange rate, emit ABCI event
+			if err = k.SetExchangeRateWithEvent(ctx, ballotDenom.Denom, exchangeRate); err != nil {
+				return err
+			}
+		}
+
+		// update miss counting & slashing
+		voteTargetsLen := len(voteTargets)
+		claimSlice := types.ClaimMapToSlice(validatorClaimMap)
+		for _, claim := range claimSlice {
+			// Skip valid voters
+			if int(claim.WinCount) == voteTargetsLen {
```

Comment:
this line needs to be >=, otherwise if you get >67% VP voting for a non-whitelisted asset then everyone gets a miss